### PR TITLE
Add relation error

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following changes in state allow us to see if a relation is dying
whilst it already exists. This way we can inform users of juju when
attempting to add a relation that already exists, but failed to do so
because it failed to remove originally.

**Note**: A subsequent PR will fix the issue where we double wrap
the `already exists (already exists):` in the error message.

## QA steps

This requires forcing a charm into a broken state on departing, so
we can see the error message.

Firstly we need to apply the diff to juju-qa-test charm:

```diff
diff --git a/testcharms/charm-hub/charms/juju-qa-test/metadata.yaml b/testcharms/charm-hub/charms/juju-qa-test/metadata.yaml
index 5f485e47d5..6373a29faf 100644
--- a/testcharms/charm-hub/charms/juju-qa-test/metadata.yaml
+++ b/testcharms/charm-hub/charms/juju-qa-test/metadata.yaml
@@ -8,6 +8,7 @@ description: |
   A container-based V2 metadata charm to use in testing juju.
   It has config, actions, resources, and a relation.
 series: [groovy, focal, bionic, xenial]
+subordinate: true
 requires:
   info:
     interface: juju-info
diff --git a/testcharms/charm-hub/charms/juju-qa-test/src/charm.py b/testcharms/charm-hub/charms/juju-qa-test/src/charm.py
index 25d36077c8..39b1d6141d 100755
--- a/testcharms/charm-hub/charms/juju-qa-test/src/charm.py
+++ b/testcharms/charm-hub/charms/juju-qa-test/src/charm.py
@@ -25,6 +25,7 @@ class ApitestUbuntuQaCharm(CharmBase):
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.fortune_action, self._on_fortune_action)
         self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.info_relation_departed, self._on_relation_departed)
         self._stored.set_default(things=[])
         self._stored.set_default(foo="")
         self._stored.set_default(status="")
@@ -76,6 +77,9 @@ class ApitestUbuntuQaCharm(CharmBase):
         status = "resource line one: {}".format( line.strip())
         self.unit.status = ActiveStatus(status)

+    def _on_relation_departed(self, event):
+        event.fail("FAIL")
+

 if __name__ == "__main__":
     main(ApitestUbuntuQaCharm)
```

Then we need to build that charm so we can deploy it later

```sh
cd testcharms/charm-hub/charms/juju-qa-test/
charmcraft build
cd -
```

```sh
$ juju bootstrap lxd test
$ juju deploy ubuntu
$ juju deploy ./testcharms/charm-hub/charms/juju-qa-test/juju-qa-test.charm
$ juju relate ubuntu juju-qa-test
$ juju remove-relation ubuntu juju-qa-test
$ juju relate ubuntu juju-qa-test
ERROR cannot add relation "juju-qa-test:info ubuntu:juju-info"
relation juju-qa-test:info ubuntu:juju-info is dying and already exists (already exists):

Use 'juju status --relations' to view the current relations.
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1902544

### CODE REVIEW

**Comment by jameinel** [comment]
The QA steps look very good though the final message:
```Use 'juju status --relations' to view the current relations
for "ubuntu, juju-qa-test"```

feels a little awkward.

**Comment by SimonRichardson** [comment]
> The QA steps look very good though the final message:
> 
> ```
> for "ubuntu, juju-qa-test"```
> 
> feels a little awkward.
> ```

Agreed, cleaned that up.

File: state/state.go [hunk 1]
```diff
@@ -2021,8 +2024,14 @@ func (st *State) AddRelation(eps ...Endpoint) (r *Relation, err error) {
 		if exists, err := isNotDead(st, relationsC, key); err != nil {
 			return nil, errors.Trace(err)
 		} else if exists {
+			// Ignore the error here, if there is an error, we know that dying
+			// will be false and can fall through to error message below.
+			if dying, _ := isDying(st, relationsC, key); dying {
+				return nil, errors.AlreadyExistsf("relation %v is dying and", key)
```
Comment (manadart) [inline state/state.go:2030]: I think something like:
```go
errors.NewAlreadyExists(nil, fmt.Sprintf("relation %v is dying, but not yet removed", key))
```
although it not as pleasing in code, results in a better error message:
- We indicate the state better without burdening the user with the specific knowledge of the `life` concept.
- We are not being tautological with "already exists".

**Review by manadart — APPROVED** [review]
I think we should edit the message for taste, but otherwise looks/works fine.

File: state/state.go (hunk 1 above)
Comment (SimonRichardson) [reply state/state.go]: Always forget about generating the error without the suffix using New!

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**